core/store/src/trie/trie_storage.rs

```diff
@@ -9,34 +9,144 @@ use crate::trie::POISONED_LOCK_ERR;
+        match self.chunk_cache.get(key) {
+            Some(value) => CachePosition::ChunkCache(value.clone()),
+            None => match self.shard_cache.get(key) {
+                Some(value) => CachePosition::ShardCache(value.clone()),
+                None => CachePosition::None,
+            },
+        }
+    }
+
+    pub fn get_with_cost(&mut self, key: &CryptoHash) -> RawBytesWithCost {
+        match self.get_cache_position(key) {
+            CachePosition::None => {
+                RawBytesWithCost { value: None, cost: TrieNodeRetrievalCost::Full }
+            }
+            CachePosition::ShardCache(value) => {
+                if let TrieCacheState::CachingChunk = self.cache_state {
+                    let value = self
+                        .shard_cache
+                        .pop(key)
+                        .expect("If position is ShardCache then value must be presented");
+                    self.chunk_cache.insert(key.clone(), value);
+                };
+                RawBytesWithCost { value: Some(value), cost: TrieNodeRetrievalCost::Full }
+            }
+            CachePosition::ChunkCache(value) => RawBytesWithCost {
+                value: Some(value),
+                cost: match self.cache_state {
+                    TrieCacheState::CachingShard => TrieNodeRetrievalCost::Full,
+                    TrieCacheState::CachingChunk => TrieNodeRetrievalCost::Free,
+                },
+            },
+        }
+    }
+
+    pub fn put(&mut self, key: CryptoHash, value: &[u8]) {
+        if value.len() >= TRIE_LIMIT_CACHED_VALUE_SIZE {
+            return;
+        }
+        let value = value.to_vec();
+
+        if let TrieCacheState::CachingChunk = &self.cache_state {
+            self.shard_cache.pop(&key);
+            self.chunk_cache.insert(key, value);
+        } else {
+            if self.chunk_cache.contains_key(&key) {
+                self.chunk_cache.insert(key, value);
+            } else {
+                self.shard_cache.put(key, value);
+            }
+        }
+    }
+
+    pub fn pop(&mut self, hash: &CryptoHash) -> Option<Vec<u8>> {
+        match self.chunk_cache.remove(hash) {
+            Some(value) => Some(value),
+            None => self.shard_cache.pop(hash),
+        }
+    }
+}
+
+impl SyncTrieCache {
     pub fn new() -> Self {
-        Self(Arc::new(Mutex::new(LruCache::new(TRIE_MAX_CACHE_SIZE))))
+        Self(Arc::new(Mutex::new(TrieCache::new(TRIE_MAX_SHARD_CACHE_SIZE))))
     }
 
     pub fn clear(&self) {
-        self.0.lock().expect(POISONED_LOCK_ERR).clear()
+        let mut guard = self.0.lock().expect(POISONED_LOCK_ERR);
+        guard.cache_state = TrieCacheState::CachingShard;
+        guard.shard_cache.clear();
+        guard.chunk_cache.clear();
+    }
+
+    pub fn set_state(&self, state: TrieCacheState) {
+        let mut guard = self.0.lock().expect(POISONED_LOCK_ERR);
+        guard.cache_state = state;
```

Comment:
and where do we have the code that "moves" stuff from chunk_cache to shard_cache when state changes ?